src/surrogate_heat_driver.cpp

```diff
@@ -87,40 +180,249 @@ void SurrogateHeatDriver::generate_arrays()
+  if (verbosity_ >= verbose::LOW) {
+    bool mass_conserved = is_mass_conserved(rho, u);
+    bool energy_conserved = is_energy_conserved(rho, u, h, channel_powers);
+
+    Expects(mass_conserved);
+    Expects(energy_conserved);
+  }
+}
+
+bool SurrogateHeatDriver::is_mass_conserved(const xt::xtensor<double, 2>& rho, const xt::xtensor<double, 2>& u) const
+{
+  bool mass_conserved = true;
+
+  for (gsl::index axial = 0; axial < n_axial_; ++axial) {
+    double mass_flowrate = 0.0;
+    for (gsl::index chan = 0; chan < n_channels_; ++ chan) {
+      double u_cell_centered = 0.5 * (u(chan, axial) + u(chan, axial + 1));
+      mass_flowrate += u_cell_centered * channels_[chan].area_ * rho(chan, axial);
+    }
+
+    double tol = std::abs(mass_flowrate - mass_flowrate_) / mass_flowrate_;
+
+    if (tol > 1e-3) {
+      mass_conserved = false;
+    }
+
+    if (verbosity_ == verbose::HIGH) {
+      std::cout << "Mass on plane " << axial << " conserved to a tolerance of " << tol << std::endl;
+    }
+  }
+
+  return mass_conserved;
+}
+
+bool SurrogateHeatDriver::is_energy_conserved(const xt::xtensor<double, 2>& rho, const xt::xtensor<double, 2>& u,
+  const xt::xtensor<double, 2>& h, const xt::xtensor<double, 2>& q) const
+{
+  bool energy_conserved = true;
+
+  for (gsl::index axial = 0; axial < n_axial_; ++axial) {
+    for (gsl::index chan = 0; chan < n_channels_; ++chan) {
+      double u_cell_centered = 0.5 * (u(chan, axial) + u(chan, axial + 1));
+      double mass_flowrate = rho(chan, axial) * channels_[chan].area_ * u_cell_centered;
+
+      // conversion factor of 1e3 to convert enthalpy from kJ/kg to J/kg
+      double channel_energy_change = mass_flowrate * (h(chan, axial + 1) - h(chan, axial)) * 1.0e3;
+
+      double tol = std::abs(channel_energy_change - q(chan, axial)) / q(chan, axial);
+
+      if (tol > 1e-3) {
+        energy_conserved = false;
+      }
+
+      if (verbosity_ == verbose::HIGH) {
+        std::cout << "Energy deposition in channel " << chan << ", axial node " << axial <<
+          " conserved to a tolerance of " << tol << std::endl;
+      }
+    }
+  }
+
+  return energy_conserved;
+}
+
+void SurrogateHeatDriver::solve_heat()
+{
+  std::cout << "Solving heat equation...\n";
+
   // Convert source to [W/m^3] as expected by Magnolia
   xt::xtensor<double, 3> q = 1e6 * source_;
 
   // Convert radial grids to [m] as expected by Magnolia
   xt::xtensor<double, 1> r_fuel = 0.01 * r_grid_fuel_;
   xt::xtensor<double, 1> r_clad = 0.01 * r_grid_clad_;
 
-  for (int i = 0; i < n_pins_; ++i) {
-    for (int j = 0; j < n_axial_; ++j) {
+  for (gsl::index i = 0; i < n_pins_; ++i) {
+    for (gsl::index j = 0; j < n_axial_; ++j) {
+      // approximate cladding surface temperature as equal to the fluid
+      // temperature, i.e. this neglects any heat transfer resistance
+      double T_co = fluid_temperature_(i, j);
+
+      // Set initial temperature to surface temperature
+      temperature_.fill(T_co);
```

Comment:
Just realized that this causes problems because it overwrites the solution for previous pins/axial segments. The following should work:
```C++
xt::view(temperature_, i, j) = T_co;
```
You'll also need `#include "xtensor/xview.hpp"` up top.

Reply:
Good catch!